File: sdk/csharp/src/Core/Protection.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;
using System.Threading;

namespace Tusk.Protection
{
    /// <summary>
    /// TuskLang SDK Protection Core Module
    /// Enterprise-grade protection for C# SDK
    /// </summary>
    public class TuskProtection
    {
        private readonly string _licenseKey;
        private readonly string _apiKey;
        private readonly string _sessionId;
        private readonly byte[] _encryptionKey;
        private readonly ConcurrentDictionary<string, string> _integrityChecks;
        private readonly UsageMetrics _usageMetrics;
        private readonly object _metricsLock = new object();

        public TuskProtection(string licenseKey, string apiKey)
        {
            _licenseKey = licenseKey ?? throw new ArgumentNullException(nameof(licenseKey));
            _apiKey = apiKey ?? throw new ArgumentNullException(nameof(apiKey));
            _sessionId = Guid.NewGuid().ToString();
            _encryptionKey = DeriveKey(licenseKey);
            _integrityChecks = new ConcurrentDictionary<string, string>();
            _usageMetrics = new UsageMetrics();
        }

        private static byte[] DeriveKey(string password)
        {
            var salt = Encoding.UTF8.GetBytes("tusklang_protection_salt");
            using var pbkdf2 = new Rfc2898DeriveBytes(password, salt, 100000, HashAlgorithmName.SHA256);
            return pbkdf2.GetBytes(32);
        }

        public bool ValidateLicense()
        {
            try
            {
                if (string.IsNullOrEmpty(_licenseKey) || _licenseKey.Length < 32)
                    return false;

                using var sha256 = SHA256.Create();
                var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(_licenseKey));
                var checksum = Convert.ToHexString(hash).ToLower();
                return checksum.StartsWith("tusk");
            }
            catch
            {
                return false;
            }
        }

        public string EncryptData(string data)
        {
            try
            {
                using var aes = Aes.Create();
                aes.Key = _encryptionKey;
                aes.GenerateIV();

                using var encryptor = aes.CreateEncryptor();
                var plaintext = Encoding.UTF8.GetBytes(data);
                var ciphertext = encryptor.TransformFinalBlock(plaintext, 0, plaintext.Length);

                var result = new byte[aes.IV.Length + ciphertext.Length];
                Buffer.BlockCopy(aes.IV, 0, result, 0, aes.IV.Length);
                Buffer.BlockCopy(ciphertext, 0, result, aes.IV.Length, ciphertext.Length);

                return Convert.ToBase64String(result);
            }
            catch
            {
                return data;
            }
        }

        public string DecryptData(string encryptedData)
        {
            try
            {
                var encryptedBytes = Convert.FromBase64String(encryptedData);
                if (encryptedBytes.Length < 16)
                    return encryptedData;

                using var aes = Aes.Create();
                aes.Key = _encryptionKey;

                var iv = new byte[16];
                var ciphertext = new byte[encryptedBytes.Length - 16];
                Buffer.BlockCopy(encryptedBytes, 0, iv, 0, 16);
                Buffer.BlockCopy(encryptedBytes, 16, ciphertext, 0, ciphertext.Length);

                aes.IV = iv;
                using var decryptor = aes.CreateDecryptor();
                var plaintext = decryptor.TransformFinalBlock(ciphertext, 0, ciphertext.Length);
                return Encoding.UTF8.GetString(plaintext);
            }
            catch
            {
                return encryptedData;
            }
        }

        public bool VerifyIntegrity(string data, string signature)
        {
            try
            {
                var expectedSignature = GenerateSignature(data);
                return string.Equals(signature, expectedSignature, StringComparison.OrdinalIgnoreCase);
            }
            catch
            {
                return false;
            }
        }

        public string GenerateSignature(string data)
        {
            try
            {
                using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(_apiKey));
                var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(data));
                return Convert.ToHexString(hash).ToLower();
            }
            catch
            {
                return string.Empty;
            }
        }

        public void TrackUsage(string operation, bool success = true)
        {
            lock (_metricsLock)
            {
                _usageMetrics.ApiCalls++;
                if (!success)
                    _usageMetrics.Errors++;
            }
        }

        public Dictionary<string, object> GetMetrics()
        {
            lock (_metricsLock)
            {
                var uptime = DateTimeOffset.UtcNow.ToUnixTimeSeconds() - _usageMetrics.StartTime;
                return new Dictionary<string, object>
                {
                    ["start_time"] = _usageMetrics.StartTime,
                    ["api_calls"] = _usageMetrics.ApiCalls,
                    ["errors"] = _usageMetrics.Errors,
                    ["session_id"] = _sessionId,
                    ["uptime"] = uptime
                };
            }
        }

        public string ObfuscateCode(string code)
        {
            var bytes = Encoding.UTF8.GetBytes(code);
            return Convert.ToBase64String(bytes);
        }

        public bool DetectTampering()
        {
            try
            {
                // In production, implement file integrity checks
                // For now, return true as placeholder
                return true;
            }
            catch
            {
                return false;
            }
        }

        public Violation ReportViolation(string violationType, string details)
        {
            var violation = new Violation
            {
                Timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                SessionId = _sessionId,
                ViolationType = violationType,
                Details = details,
                LicenseKeyPartial = _licenseKey.Length > 8 
                    ? _licenseKey.Substring(0, 8) + "..." 
                    : _licenseKey
            };

            Console.WriteLine($"SECURITY VIOLATION: {violation}");
            return violation;
        }

        // Inner classes
        public class UsageMetrics
        {
            public long StartTime { get; set; }
            public long ApiCalls { get; set; }
            public long Errors { get; set; }

            public UsageMetrics()
            {
                StartTime = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
                ApiCalls = 0;
                Errors = 0;
            }
        }

        public class Violation
        {
            public long Timestamp { get; set; }
            public string SessionId { get; set; }
            public string ViolationType { get; set; }
            public string Details { get; set; }
            public string LicenseKeyPartial { get; set; }

            public override string ToString()
            {
                return $"Violation{{Timestamp={Timestamp}, SessionId='{SessionId}', Type='{ViolationType}', Details='{Details}', LicenseKeyPartial='{LicenseKeyPartial}'}}";
            }
        }

        // Global protection instance
        private static volatile TuskProtection _protectionInstance;
        private static readonly object _instanceLock = new object();

        public static TuskProtection InitializeProtection(string licenseKey, string apiKey)
        {
            if (_protectionInstance == null)
            {
                lock (_instanceLock)
                {
                    if (_protectionInstance == null)
                    {
                        _protectionInstance = new TuskProtection(licenseKey, apiKey);
                    }
                }
            }
            return _protectionInstance;
        }

        public static TuskProtection GetProtection()
        {
            if (_protectionInstance == null)
            {
                throw new InvalidOperationException("Protection not initialized. Call InitializeProtection() first.");
            }
            return _protectionInstance;
        }
    }
} 

File: sdk/javascript/build-fixed/tusklang-javascript-sdk/core/src/Protection.js
/**
 * TuskLang SDK Protection Core Module
 * Enterprise-grade protection for JavaScript SDK
 */

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

class UsageMetrics {
    constructor() {
        this.startTime = Date.now();
        this.apiCalls = 0;
        this.errors = 0;
    }

    incrementApiCalls() {
        this.apiCalls++;
    }

    incrementErrors() {
        this.errors++;
    }

    getStartTime() {
        return this.startTime;
    }

    getApiCalls() {
        return this.apiCalls;
    }

    getErrors() {
        return this.errors;
    }
}

class Violation {
    constructor(timestamp, sessionId, violationType, details, licenseKeyPartial) {
        this.timestamp = timestamp;
        this.sessionId = sessionId;
        this.violationType = violationType;
        this.details = details;
        this.licenseKeyPartial = licenseKeyPartial;
    }

    toString() {
        return `VIOLATION [${new Date(this.timestamp).toISOString()}] ${this.violationType}: ${this.details} (Session: ${this.sessionId}, License: ${this.licenseKeyPartial})`;
    }
}

class TuskProtection {
    constructor(licenseKey, apiKey) {
        this.licenseKey = licenseKey;
        this.apiKey = apiKey;
        this.sessionId = this.generateUUID();
        this.encryptionKey = this.deriveKey(licenseKey);
        this.integrityChecks = [];
        this.usageMetrics = new UsageMetrics();
    }

    generateUUID() {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
            const r = Math.random() * 16 | 0;
            const v = c == 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    }

    deriveKey(password) {
        const salt = 'tusklang_protection_salt';
        return crypto.pbkdf2Sync(password, salt, 100000, 32, 'sha256');
    }

    validateLicense() {
        try {
            if (this.licenseKey.length < 32) {
                return false;
            }

            const checksum = crypto.createHash('sha256').update(this.licenseKey).digest('hex');
            return checksum.startsWith('tusk');
        } catch (error) {
            return false;
        }
    }

    encryptData(data) {
        try {
            const iv = crypto.randomBytes(12);
            const cipher = crypto.createCipherGCM('aes-256-gcm', this.encryptionKey);
            
            let encrypted = cipher.update(data, 'utf8', 'hex');
            encrypted += cipher.final('hex');
            
            const tag = cipher.getAuthTag();
            
            const result = Buffer.concat([iv, tag, Buffer.from(encrypted, 'hex')]);
            return result.toString('base64');
        } catch (error) {
            return data;
        }
    }

    decryptData(encryptedData) {
        try {
            const decoded = Buffer.from(encryptedData, 'base64');
            if (decoded.length < 28) {
                return encryptedData;
            }

            const iv = decoded.slice(0, 12);
            const tag = decoded.slice(12, 28);
            const encrypted = decoded.slice(28);

            const decipher = crypto.createDecipherGCM('aes-256-gcm', this.encryptionKey);
            decipher.setAuthTag(tag);
            
            let decrypted = decipher.update(encrypted, null, 'utf8');
            decrypted += decipher.final('utf8');
            
            return decrypted;
        } catch (error) {
            return encryptedData;
        }
    }

    verifyIntegrity(data, signature) {
        try {
            const expectedSignature = this.generateSignature(data);
            return crypto.timingSafeEqual(Buffer.from(signature), Buffer.from(expectedSignature));
        } catch (error) {
            return false;
        }
    }

    generateSignature(data) {
        return crypto.createHmac('sha256', this.apiKey).update(data).digest('hex');
    }

    trackUsage(operation, success = true) {
        this.usageMetrics.incrementApiCalls();
        if (!success) {
            this.usageMetrics.incrementErrors();
        }
    }

    getMetrics() {
        return {
            start_time: this.usageMetrics.getStartTime(),
            api_calls: this.usageMetrics.getApiCalls(),
            errors: this.usageMetrics.getErrors(),
            session_id: this.sessionId,
            uptime: Date.now() - this.usageMetrics.getStartTime()
        };
    }

    obfuscateCode(code) {
        return Buffer.from(code).toString('base64');
    }

    detectTampering() {
        try {
            // In production, implement file integrity checks
            // For now, return true as placeholder
            return true;
        } catch (error) {
            return false;
        }
    }

    reportViolation(violationType, details) {
        const violation = new Violation(
            Date.now(),
            this.sessionId,
            violationType,
            details,
            this.licenseKey.substring(0, 8) + '...'
        );

        console.error("SECURITY VIOLATION: " + violation.toString());
        return violation;
    }

    // Advanced security features
    encryptFile(filePath) {
        try {
            const content = fs.readFileSync(filePath, 'utf8');
            const encrypted = this.encryptData(content);
            const encryptedPath = filePath + '.enc';
            fs.writeFileSync(encryptedPath, encrypted);
            return encryptedPath;
        } catch (error) {
            throw new Error(`Failed to encrypt file: ${error.message}`);
        }
    }

    decryptFile(encryptedFilePath) {
        try {
            const encrypted = fs.readFileSync(encryptedFilePath, 'utf8');
            const decrypted = this.decryptData(encrypted);
            const decryptedPath = encryptedFilePath.replace('.enc', '');
            fs.writeFileSync(decryptedPath, decrypted);
            return decryptedPath;
        } catch (error) {
            throw new Error(`Failed to decrypt file: ${error.message}`);
        }
    }

    generateSecureToken() {
        return crypto.randomBytes(32).toString('hex');
    }

    hashPassword(password) {
        const salt = crypto.randomBytes(16).toString('hex');
        const hash = crypto.pbkdf2Sync(password, salt, 100000, 64, 'sha512').toString('hex');
        return `${salt}:${hash}`;
    }

    verifyPassword(password, hashedPassword) {
        const [salt, hash] = hashedPassword.split(':');
        const verifyHash = crypto.pbkdf2Sync(password, salt, 100000, 64, 'sha512').toString('hex');
        return crypto.timingSafeEqual(Buffer.from(hash), Buffer.from(verifyHash));
    }

    // Self-destruct mechanism
    selfDestruct() {
        try {
            // Clear sensitive data
            this.encryptionKey = null;
            this.licenseKey = null;
            this.apiKey = null;
            
            // Report violation
            this.reportViolation('SELF_DESTRUCT', 'Protection system activated self-destruct');
            
            // In production, this could trigger additional cleanup
            return true;
        } catch (error) {
            return false;
        }
    }
}

// Singleton pattern
let protectionInstance = null;

function initializeProtection(licenseKey, apiKey) {
    if (!protectionInstance) {
        protectionInstance = new TuskProtection(licenseKey, apiKey);
    }
    return protectionInstance;
}

function getProtection() {
    if (!protectionInstance) {
        throw new Error('Protection not initialized. Call initializeProtection() first.');
    }
    return protectionInstance;
}

module.exports = {
    TuskProtection,
    UsageMetrics,
    Violation,
    initializeProtection,
    getProtection
}; 